server/src/test/java/org/elasticsearch/cluster/coordination/PublicationTests.java

```diff
@@ -0,0 +1,362 @@
+    MockNode node3 = new MockNode(Settings.EMPTY, n3);
+    List<MockNode> nodes = Arrays.asList(node1, node2, node3);
+
+    Function<DiscoveryNode, MockNode> nodeResolver = dn -> nodes.stream().filter(mn -> mn.localNode.equals(dn)).findFirst().get();
+
+    private void initializeCluster(VotingConfiguration initialConfig) {
+        node1.coordinationState.setInitialState(CoordinationStateTests.clusterState(0L, 1L, n1, initialConfig, initialConfig, 0L));
+        StartJoinRequest startJoinRequest = new StartJoinRequest(n1, 1L);
+        node1.coordinationState.handleJoin(node1.coordinationState.handleStartJoin(startJoinRequest));
+        node1.coordinationState.handleJoin(node2.coordinationState.handleStartJoin(startJoinRequest));
+        node1.coordinationState.handleJoin(node3.coordinationState.handleStartJoin(startJoinRequest));
+        assertTrue(node1.coordinationState.electionWon());
+    }
+
+    public void testSimpleClusterStatePublishing() throws InterruptedException {
+        VotingConfiguration singleNodeConfig = new VotingConfiguration(Sets.newHashSet(n1.getId()));
+        initializeCluster(singleNodeConfig);
+
+        AssertingAckListener ackListener = new AssertingAckListener(nodes.size());
+        DiscoveryNodes discoveryNodes = DiscoveryNodes.builder().add(n1).add(n2).add(n3).localNodeId(n1.getId()).build();
+        MockPublication publication = node1.publish(CoordinationStateTests.clusterState(1L, 2L,
+            discoveryNodes, singleNodeConfig, singleNodeConfig, 42L), ackListener, Collections.emptySet());
+
+        assertThat(publication.pendingPublications.keySet(), equalTo(discoNodes));
+        assertTrue(publication.pendingCommits.isEmpty());
+        publication.pendingPublications.entrySet().stream().collect(shuffle()).forEach(e -> {
+            PublishResponse publishResponse = nodeResolver.apply(e.getKey()).coordinationState.handlePublishRequest(
+                publication.publishRequest);
+            e.getValue().onResponse(new PublishWithJoinResponse(publishResponse, Optional.empty()));
+        });
+
+        assertThat(publication.pendingCommits.keySet(), equalTo(discoNodes));
+        assertNotNull(publication.applyCommit);
+        assertEquals(publication.applyCommit.getTerm(), publication.publishRequest.getAcceptedState().term());
+        assertEquals(publication.applyCommit.getVersion(), publication.publishRequest.getAcceptedState().version());
+        publication.pendingCommits.entrySet().stream().collect(shuffle()).forEach(e -> {
+            nodeResolver.apply(e.getKey()).coordinationState.handleCommit(publication.applyCommit);
+            e.getValue().onResponse(TransportResponse.Empty.INSTANCE);
+        });
+
+        assertTrue(publication.completed);
+        assertTrue(publication.committed);
+
+        assertThat(ackListener.await(0L, TimeUnit.SECONDS).size(), equalTo(3));
+    }
+
+    public void testClusterStatePublishingWithFaultyNode() throws InterruptedException {
```

Comment:
I think that this test is trying to test too many things via its randomly-set flags, making it hard to tell what it's actually covering. Please could we split out the different cases into separate tests a bit?

Reply:
I've split this into two tests in a76c4f1c7cd